[hansneddyanto] iP

hansneddyanto opened the pull request:
# Hana

> “Your mind is for having ideas, not holding them.” – David Allen [(source)](https://dansilvestre.com/productivity-quotes/)

Hana frees your mind of having to remember things you need to do. It's:

- text-based
- easy to learn
- ~~FAST~~ **SUPER FAST** to use

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2425s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😊

And it is **FREE**!

## Features:
- [x] Managing tasks
- [x] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

---

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

hansneddyanto committed:
Rename Duke to Hana
Add greet and exit features

hansneddyanto committed:
Make minor modifications to code structure
Adds echo functionality

hansneddyanto committed:
Add the features to add and list tasks

hansneddyanto committed:
Add features to mark and unmark task

hansneddyanto committed:
Add features for ToDos, Events, and Deadlines

hansneddyanto committed:
Add automated text UI testing

hansneddyanto committed:
Handle errors (without HanaException class for now)

hansneddyanto committed:
Add feature to delete task

hansneddyanto committed:
Handle errors using HanaException class

hansneddyanto committed:
Add feature to save tasks in hard disk (hana.txt) using class Storage

hansneddyanto committed:
Merge branch 'branch-Level-7'

hansneddyanto committed:
Add feature to understand date. Time input needs to be in the format yyyy-mm-dd

hansneddyanto committed:
Merge branch 'branch-Level-8'

hansneddyanto committed:
Use more OOP by adding classes Parser, TaskList, Ui, and Command. Edit Storage class too.

hansneddyanto committed:
Divide classes into packages (hana and hana.task)

hansneddyanto committed:
Make minor changes to hana.txt

hansneddyanto committed:
Merge branch 'add-gradle-support'

hansneddyanto committed:
Add Gradle

hansneddyanto committed:
Add JUnit for ParserTest and TaskListTest

hansneddyanto committed:
Add JAR

hansneddyanto committed:
Put JavaDoc comments on all classes and methods in the files

hansneddyanto committed:
Adhere to coding standard. Changed some variable names to more meaningful ones as recommended

hansneddyanto committed:
Merge branch 'branch-A-CodingStandard'

hansneddyanto committed:
Put more JavaDoc comments to the other files

hansneddyanto committed:
Merge branch 'branch-A-JavaDoc'

hansneddyanto committed:
Add feature to find based on search word

hansneddyanto committed:
Merge branch 'branch-Level-9'

lrongyi commented on src/main/java/hana/Parser.java:
```diff
@@ -0,0 +1,108 @@
+        case "list":
+            return new ListCommand();
+        case "mark":
+            Pattern markPattern = Pattern.compile("^mark (\\d+)$");
```
Good use of descriptive names.

lrongyi commented on src/main/java/hana/Ui.java:
```diff
@@ -0,0 +1,114 @@
+ * Handles interactions with the user.
+ */
+public class Ui {
+    private final String line = "    ____________________________________________________________\n";
```
Final variables should be in all caps.

lrongyi commented on src/main/java/hana/Command.java:
```diff
@@ -0,0 +1,121 @@
+}
+
+class AddCommand extends Command {
+    private final Task task;
```
Final variables should be in all caps.

lrongyi commented on src/main/java/hana/Command.java:
```diff
@@ -0,0 +1,121 @@
+}
+
+class DeleteCommand extends Command {
+    private final int index;
```
Final variables in all caps.

lrongyi commented on src/main/java/hana/Command.java:
```diff
@@ -0,0 +1,121 @@
+}
+
+class MarkCommand extends Command {
+    private final int index;
```
Final variables in all caps.

lrongyi commented on src/main/java/hana/Command.java:
```diff
@@ -0,0 +1,121 @@
+}
+
+class UnmarkCommand extends Command {
+    private final int index;
```
Final variables in all caps.

lrongyi reviewed (commented):
LTGM overall. No issues in terms of code quality, only issue with coding standard is the naming of final variables.

jagdeepsh commented on src/main/java/hana/Hana.java:
```diff
@@ -0,0 +1,60 @@
+
+    private Storage storage;
+    private TaskList tasks;
+    private Ui ui;
```
Good use of naming practice as per coding standard

jagdeepsh commented on src/main/java/hana/Parser.java:
```diff
@@ -0,0 +1,108 @@
+        switch (commandWord) {
+        case "list":
+            return new ListCommand();
+        case "mark":
```
Good job not indenting `switch` statement

jagdeepsh commented on src/main/java/hana/Command.java:
```diff
@@ -0,0 +1,121 @@
+}
+
+class FindCommand extends Command {
+    private String keyword;
```
Consider using `String command` instead of `String keyword`?

jagdeepsh commented on src/main/java/hana/Storage.java:
```diff
@@ -0,0 +1,111 @@
+     * @return A string representing the Task.
+     */
+    private String taskToString(Task task) {
+        if (task instanceof ToDo) {
+            return "T | " + (task.isDone ? "1" : "0") + " | " + task.description;
+        } else if (task instanceof Deadline) {
+            Deadline d = (Deadline) task;
+            return "D | " + (task.isDone ? "1" : "0") + " | " + d.description + " | " + d.by;
+        } else if (task instanceof Event) {
+            Event e = (Event) task;
+            return "E | " + (task.isDone ? "1" : "0") + " | " + e.description + " | " + e.from + " | " + e.to;
+        }
+        return "";
```
Great application of indentation in accordance to the coding standards

jagdeepsh reviewed (commented):
Looks good, just some minor tweaks needed.

hansneddyanto committed:
Change from CLI to GUI

hansneddyanto committed:
Put some minor edits for the "bye" checking statement

hansneddyanto committed:
Edit javadocs for MainWindow.java handleUserInput() method

hansneddyanto committed:
Merge branch 'branch-Level-10'

hansneddyanto committed:
Add checkstyle and edit files according to checkstyle

hansneddyanto committed:
Merge branch 'branch-A-Checkstyle'

hansneddyanto committed:
Make minor changes to hana.txt

hansneddyanto committed:
Create jar file and handle empty description error for Event and Deadline

hansneddyanto committed:
Add assertions to document important assumptions in Parser class

Added assertions at key points in the Parser class to ensure important
assumptions hold during runtime. These assertions help catch logical
issues during development by validating the state of variables before
further operations.

- Validate that task numbers and task indices are non-negative to avoid
  index-related issues.
- Validate that in the find method, only one keyword (can be with space)
  is available
- Ensure that event start and end dates are valid and that the start
  date is not after the end date.

These assertions will help during development when assertions are
enabled (-ea flag) and will prevent invalid state progression.

A little bit of refactoring done on Ui.java as well.

hansneddyanto committed:
Codebase contains inconsistent indentation, redundant variables, and non-uniform variable names

The current code lacks uniformity in formatting and variable naming,
making it harder to maintain and less readable. Redundant variables
increase complexity without adding value, and inconsistent indentation
makes the code look disorganized.

Improve code readability by correcting indentation, removing redundant
variables, and standardizing variable names. Ensure that all methods
and classes follow a consistent naming convention and appropriate
formatting.

These changes are made to enhance maintainability, making it easier
for developers to read and work with the code. Consistent formatting
and naming conventions also reduce the cognitive load when navigating
the codebase.

This commit does not alter any functionality but focuses solely on
refining code structure and quality.

hansneddyanto committed:
Merge pull request #1 from hansneddyanto/branch-A-Assertions

Add assertions to document important assumptions in Parser class

hansneddyanto committed:
Merge branch 'master' of https://github.com/hansneddyanto/ip

* 'master' of https://github.com/hansneddyanto/ip:
  Add assertions to document important assumptions in Parser class

hansneddyanto committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/hana/Parser.java
#	src/main/java/hana/Ui.java

hansneddyanto committed:
Merge pull request #2 from hansneddyanto/branch-A-CodeQuality

Improve on Code Quality

hansneddyanto committed:
Set up Continuous Integration using GitHub Actions

The project currently lacks automated testing and continuous
integration (CI), which makes it harder to ensure that new changes
don’t break existing functionality. Having CI will allow the project
to automatically run tests on each push or pull request, providing
quick feedback on code quality and potential issues.

Set up GitHub Actions for CI by adding a workflow configuration in a
`.yml` file. This workflow automatically builds the project and runs
tests on every commit and pull request.

This CI setup ensures that all commits are automatically verified,
increasing code stability and reducing the risk of introducing
breaking changes. GitHub Actions was chosen for its seamless
integration with GitHub repositories.

hansneddyanto committed:
Merge branch 'master' of https://github.com/hansneddyanto/ip

* 'master' of https://github.com/hansneddyanto/ip:
  Codebase contains inconsistent indentation, redundant variables, and non-uniform variable names

hansneddyanto committed:
Refine CI using GitHub Actions

The project fails automated testing and CI due to
last three statements in gradle.yml file.

The last three segments of the workflow related to
I/O redirection tasks were removed, as they are not
applicable to this project.

This CI setup ensures that all commits are automatically
verified. GitHub Actions was chosen for its seamless
integration with GitHub repositories.

hansneddyanto committed:
Add support for mass operations like mark, unmark, and delete tasks

The current system requires users to perform task operations one at a
time, making it inefficient to manage a large number of tasks.

Allowing users to perform operations on multiple tasks in one go
improves the efficiency of task management, especially when dealing
with large lists.

Implement mass operations by adding `MassMarkCommand`, `MassUnmarkCommand`,
and `MassDeleteCommand` to handle bulk task actions. These commands
search for tasks based on a keyword, display the matching tasks, and
prompt the user for confirmation before marking, unmarking, or deleting
tasks in bulk.

This approach ensures that users have control over bulk actions through
a confirmation process, reducing the chance of accidental changes.
It also maintains consistency with the existing command processing
flow, ensuring the user experience remains intuitive.

Multi-step user interactions are handled by storing the command state
until the user confirms the action, enabling a smooth multi-step
operation for bulk task management.

hansneddyanto committed:
Improve code styling by reducing nested statements in method getResponse(String input)

hansneddyanto committed:
Refactor Parser.java by extracting methods in parse()

The parse() method in Parser.java is currently too long and contains
multiple responsibilities, making it harder to read and maintain.

This needs to change because it violates the single responsibility
principle and reduces code clarity, making future modifications more
error-prone.

Extract helper methods from the parse() method to shorten it and
improve readability. Group related code into distinct methods to
better separate concerns and follow clean code principles.

This approach is chosen to promote modularity, allowing each extracted
method to handle a specific aspect of the parsing logic. This reduces
cognitive load and enhances testability.

No functional changes are introduced, and this refactor focuses solely
on improving code structure.

hansneddyanto committed:
Add Javadoc comments for command handler methods in Parser.java

The command handler methods in Parser.java currently lack Javadoc
comments, which makes it difficult for developers to understand their
purpose, behavior, and usage.

This needs to change to improve code readability and maintainability,
and to ensure the codebase follows best practices for documentation.

Add detailed Javadoc comments for each command handler method,
providing information about parameters, return values, exceptions, and
usage.

This approach is taken to ensure the code is self-documenting, making
it easier for developers to work with the command-handling logic
without needing to examine the method internals. The added comments
enhance future maintainability and onboarding.

No functional changes have been introduced.

hansneddyanto committed:
Correct assertion mistakes in handleMark, handleUnmark, and handleDelete

The handleMark, handleUnmark, and handleDelete methods contain
incorrect assertions related to task numbers, which assume task indices
are valid before parsing or validation occurs.

This needs to change because invalid assertions can cause unexpected
behavior and crashes during runtime, especially when handling user
input.

Correct the assertions by ensuring task number validation happens after
input parsing and before assertions are made. Validate the task number
to ensure it is non-negative and within the valid range before
proceeding with the command creation.

This change improves the robustness of the command handlers by
preventing potential crashes caused by incorrect assertions, making the
code more reliable and user-friendly.

No functional changes beyond the corrected assertions have been made.

hansneddyanto committed:
Remove redundant String[] parts variable in handleDelete method

The handleDelete method in Parser.java contains a redundant String[]
parts variable, which is not necessary for task number extraction and
validation.

This needs to change to simplify the code, improve readability, and
avoid unnecessary variable usage that doesn't contribute to the
functionality.

Remove the redundant parts variable and use the Matcher to extract the
task number directly from the user input. This reduces complexity and
makes the code cleaner without affecting functionality.

This approach ensures that the code is more concise and avoids
unnecessary variable declarations, improving maintainability.

No functional changes are introduced. The code is simply optimized by
removing redundancy.

hansneddyanto committed:
Add User Guide to README.md and correct UI for showUnmarkMessage

The README.md file currently lacks a detailed user guide explaining how
to use Hana, including its various commands and features.

This needs to change to provide users with comprehensive instructions
on how to interact with Hana, ensuring they understand how to add,
list, mark, unmark, and delete tasks, as well as perform mass
operations.

Added a detailed user guide in README.md, following the required
format. It includes explanations for adding tasks, marking/unmarking
tasks, and mass operations, with examples and expected outputs.
Additionally, Ui.png was added to enhance the guide with a product
screenshot.

Also, corrected a UI mistake in showUnmarkMessage to ensure the correct
message is displayed when unmarking tasks.

These changes enhance user experience by providing clear documentation
and fixing a UI issue.

hansneddyanto committed:
Remove unused Scanner import from Ui.java

The Ui.java file currently imports the Scanner class, which is not
being used anywhere in the file.

This needs to change to clean up the code by removing unnecessary
imports, which helps maintain clarity and avoid potential confusion
in the future.

Remove the unused Scanner import statement from Ui.java, as it is
redundant and not required for the current functionality.

This approach is taken to keep the codebase clean, efficient, and
free from unused imports, adhering to best coding practices.

No functionality is affected by this change; it simply improves
code hygiene.
